Draft: Arm target docs changes

jonathanpallant opened the pull request:
DO NOT MERGE. Open a PR upstream when we're ready.

jonathanpallant committed:
Add pages for every Cortex-M target

jordens committed:
arm-none-eabi: spelling/wording

diondokter commented on src/doc/rustc/src/platform-support/thumbv7em-none-eabihf.md:
```diff
@@ -0,0 +1,31 @@
+
+See [`arm-none-eabi`](arm-none-eabi.md) for information applicable to all `arm-none-eabi` targets.
+
+This target uses the hard-float ABI: functions which take `f32` or `f64` as arguments will have them passed via FPU registers. This target therefore srequire the use of an FPU (which is optional on Cortex-M4 and Cortex-M7). See also the soft-float ABI version of this target [`thumbv7em-none-eabi`](thumbv7em-none-eabi.md).
```
```suggestion
This target uses the hard-float ABI: functions which take `f32` or `f64` as arguments will have them passed via FPU registers. This target therefore requires the use of an FPU (which is optional on Cortex-M4 and Cortex-M7). See also the soft-float ABI version of this target [`thumbv7em-none-eabi`](thumbv7em-none-eabi.md).
```

Typo

jonathanpallant committed:
Typo in thumbv7em-none-eabihf.

jonathanpallant committed:
Add more Cortex-M details

Now explains how to, for example, support a Cortex-M55 with FPU and Integer Helium.

jonathanpallant committed:
Table formatting

chrisnc commented on src/doc/rustc/src/platform-support/arm-none-eabi.md:
```diff
@@ -69,16 +140,23 @@ build-std = ["core"]
-  the CPU.
-* Architectures prior to ARMv7 don't have atomic instructions.
+  the target, `target-cpu` and `target-feature`s.
+* Architectures prior to ARMv7 don't have atomic compare-and-swap instructions
```
This bullet is incorrect in a few ways:
 - There are architectures prior to Armv7 that have atomics (like Armv6), and architectures after it that don't (like Armv6-M). The original wording just said "old enough that it doesn't have atomic instructions", but the reality is more complicated; we don't have a non-M `armv6-none` target yet, but one could exist in theory.
 - AArch64 has compare-and-swap instructions, but Arm does not. Arm has exclusive load+store, which are slightly more general than CAS, and CAS can be implemented on top of them.
 - `fetch_add` is not a compare-and-swap; it unconditionally stores a new value and returns the old value. It can be implemented in terms of a compare-and-swap, but this is not how it's implemented on Arm.

jonathanpallant replied:
Interesting. I was going off how https://docs.rs/portable-atomic phrased things. I'll fix it!

jonathanpallant replied:
https://doc.rust-lang.org/stable/std/sync/atomic/index.html#portability also talks about Compare and Swap, and not Load Exclusive.

jonathanpallant replied:
I think I fixed it

chrisnc replied:
Ha, yeah, the wording in sync::atomic's docs is misleading. I think it probably comes from our x86-dominated history, where indeed CAS is the fundamental atomic primitive that everything sits on top of.

Updated wording looks good!

jonathanpallant committed:
Merge pull request #3 from quartiq/arm-target-changes

arm-none-eabi: spelling/wording

jonathanpallant committed:
Clarify wording around atomics.

jonathanpallant committed:
Tweak FPU wording.
